Source/UserSession/ZMUserSession+Push.swift

```diff
@@ -101,28 +102,58 @@ extension ZMUserSession {
     func setPushToken(_ pushToken: PushToken) {
         let syncMOC = managedObjectContext.zm_sync!
         syncMOC.performGroupedBlock {
-            guard let selfClient = ZMUser.selfUser(in: syncMOC).selfClient() else { return }
-            if selfClient.pushToken?.deviceToken != pushToken.deviceToken {
-                selfClient.pushToken = pushToken
-                syncMOC.saveOrRollback()
+            guard
+                let selfClient = ZMUser.selfUser(in: syncMOC).selfClient(),
+                let clientID = selfClient.remoteIdentifier,
+                selfClient.pushToken?.deviceToken != pushToken.deviceToken
+            else {
+                return
+            }
+
+            let action = RegisterPushTokenAction(token: pushToken, clientID: clientID) { result in
+                switch result {
+                case .success:
+                    selfClient.pushToken = pushToken
+                    syncMOC.saveOrRollback()
+
+                case .failure(let error):
+                    Logging.push.safePublic("Failed to register push token with backend: \(error)")
+                }
             }
+
+            action.send(in: syncMOC.notificationContext)
         }
     }
 
-    func deletePushKitToken(isLegacy: Bool = false) {
+    func deletePushKitToken(_ completion: (() -> Void)? = nil) {
```

Comment:
Usually completion handlers have an argument label:

```suggestion
    func deletePushKitToken(completion: (() -> Void)? = nil) {
```